src/Runtime/ExecutionSession.cpp

```diff
@@ -34,41 +34,27 @@ ExecutionSession::ExecutionSession(
     std::string sharedLibPath, bool defaultEntryPoint) {
   _sharedLibraryHandle =
       llvm::sys::DynamicLibrary::getPermanentLibrary(sharedLibPath.c_str());
-  if (!_sharedLibraryHandle.isValid()) {
-    std::stringstream errStr;
-    errStr << "Cannot open library: '" << sharedLibPath << "'" << std::endl;
-    throw std::runtime_error(errStr.str());
-  }
+  if (!_sharedLibraryHandle.isValid())
+    throw std::runtime_error(reportLibraryOpeningError(sharedLibPath));
 
   if (defaultEntryPoint)
     setEntryPoint("run_main_graph");
 
   _queryEntryPointsFunc = reinterpret_cast<queryEntryPointsFuncType>(
       _sharedLibraryHandle.getAddressOfSymbol(_queryEntryPointsName.c_str()));
-  if (!_queryEntryPointsFunc) {
-    std::stringstream errStr;
-    errStr << "Cannot load symbol: '" << _queryEntryPointsName << "'"
-           << std::endl;
-    throw std::runtime_error(errStr.str());
-  }
+  if (!_queryEntryPointsFunc)
+    throw std::runtime_error(reportSymbolLoadingError(_queryEntryPointsName));
 
   _inputSignatureFunc = reinterpret_cast<signatureFuncType>(
       _sharedLibraryHandle.getAddressOfSymbol(_inputSignatureName.c_str()));
-  if (!_inputSignatureFunc) {
-    std::stringstream errStr;
-    errStr << "Cannot load symbol: '" << _inputSignatureName << "'"
-           << std::endl;
-    throw std::runtime_error(errStr.str());
-  }
+  if (!_inputSignatureFunc)
+    throw std::runtime_error(reportSymbolLoadingError(_inputSignatureName));
 
   _outputSignatureFunc = reinterpret_cast<signatureFuncType>(
       _sharedLibraryHandle.getAddressOfSymbol(_outputSignatureName.c_str()));
-  if (!_outputSignatureFunc) {
-    std::stringstream errStr;
-    errStr << "Cannot load symbol: '" << _outputSignatureName << "'"
-           << std::endl;
-    throw std::runtime_error(errStr.str());
-  }
+  if (!_outputSignatureFunc)
+    throw std::runtime_error(reportSymbolLoadingError(_outputSignatureName));
+  errno = 0; // No errors.
```

Comment:
Do you have to explicitly set `errno` to zero in all these cases?

Reply:
errno is only meaningful when a function returns error, i.e., return NULL or -1. So there is never a need to set errno when a function succeeds. `if (error) errno = x` is just bad code.